Add comprehensive Zod v4 support

steven-terrana opened the pull request:
## Summary

This PR adds comprehensive Zod v4 support to zod-fast-check, enabling property-based testing with the latest Zod version.

### New Features

**Separate entry points for Zod v3 and v4:**
```ts
// Zod 3 (default, backwards compatible)
import { ZodFastCheck } from "zod-fast-check";

// Zod 4
import { ZodFastCheck } from "zod-fast-check/v4";
```

**50+ new Zod v4 features supported:**

- **Schema types**: xor, prefault, nonoptional, readonly, success, codec, strictObject, looseObject, keyof
- **String formats**: ulid, nanoid, guid, xid, ksuid, ipv4, ipv6, cidrv4, cidrv6, mac, e164, base64, base64url, jwt, hex, emoji, hostname
- **ISO namespace**: z.iso.datetime(), z.iso.date(), z.iso.time(), z.iso.duration()
- **Constraints**: bigint min/max/positive/negative, date min/max, array length, tuple rest

**Updated README** - Documents Zod version support and all new features

### Test Coverage

280 tests passing (155 for Zod 3, 125 for Zod 4)

### Known Limitations

The following Zod v4 types remain unsupported due to architectural constraints:
- `z.json()` - Uses lazy internally
- `z.file()` - Browser File API type
- `z.templateLiteral()` - Complex regex generation
- `z.custom()` - User-defined validation

Fixes #26
Fixes #22

steven-terrana committed:
Add comprehensive Zod v4 support

- Add Zod4FastCheck class for Zod v4 schemas
- Implement 50+ new Zod v4 features:
  - Schema types: xor, prefault, nonoptional, readonly, success, codec
  - String formats: ulid, nanoid, guid, xid, ksuid, ipv4, ipv6, cidrv4,
    cidrv6, mac, e164, base64, base64url, jwt, hex, emoji, hostname
  - ISO namespace: datetime, date, time, duration
  - Constraints: bigint min/max, date min/max, array length, tuple rest
  - Object variants: strictObject, looseObject, keyof
- Update README with Zod version support documentation
- Add comprehensive test coverage (280 tests passing)

steven-terrana committed:
Re-export Zod4FastCheck from main entry point

Allow users to import Zod4FastCheck directly from 'zod-fast-check'
instead of 'zod-fast-check/dist/zod4-fast-check'.

historydev commented:
regex not supported?

steven-terrana committed:
Add separate entry points for Zod v3 and v4

- Export ZodFastCheck from both /v3 and /v4 entry points
- Default import maintains backwards compatibility with Zod v3
- Add build script to rewrite imports for consumer compatibility
- Update peerDependencies to support zod ^3.0.0 || ^4.0.0
- Bump version to 0.11.0

steven-terrana committed:
Add native regex support using fc.stringMatching() for Zod v4

- Add hasUnsupportedRegexFeatures() helper to detect regex patterns that
  fc.stringMatching() cannot handle (word boundaries, lookahead/lookbehind)
- Modify string handler to extract regex patterns from Zod v4 checks and
  generate matching strings efficiently using fc.stringMatching()
- Fall back to filtering for unsupported regex features
- Add comprehensive tests for regex support including length constraints,
  startsWith/endsWith combinations, and complex patterns
- Update README with documentation on supported/unsupported regex features

steven-terrana commented:
good catch @historydev - i have updated the PR to include better support for regex.
there are some inherent limitations, which are now documented in the `README.md`

historydev commented:
**test.ts**:
```javascript
import * as fc from 'fast-check';
import {ZodFastCheck as zfc} from 'zod-fast-check/src/zod4-fast-check.ts';
import {z} from 'zod';

const User = z.object({
    firstname: z.string()
});

const userArb = zfc()
    .inputOf(User);

await fc.assert(
    fc.asyncProperty(userArb, async (input) =>
        console.log(input)
    ),
    {numRuns: 1}
);
```



**package.json/devDeps**:
```javascript
"zod-fast-check": "github:steven-terrana/zod-fast-check#zod-v4"
```

**throw**:
```javascript
error: Cannot find module 'zod-fast-check/src/zod4-fast-check.ts'
```

But this for commit 387468549acde501a24f3fa68613e33d7910bc60 (before regex) works:
```javascript
import {Zod4FastCheck as zfc} from 'zod-fast-check/src/zod4-fast-check';
```

historydev commented:
min, max, length ask override:
` Unable to generate valid values for Zod schema. An override is must be provided for the schema at path '.field'.`

steven-terrana committed:
Add tests for string constraints in nested schemas

- Add diagnostic tests verifying Zod 4 stores min/max/length in _zod.bag
- Add nested schema tests for string constraints (PR #27 issue)
- Fix jest/tsconfig to properly resolve zod4/zod3 module aliases

historydev commented:
How can I help you? I really need it. To start, I can provide a schema from my project for testing:

```javascript
const userSchema = z.object({
    id: z.int(),

    roles: z.array(z.int()),

    firstname: z.string().min(1).max(50).regex(/^[a-zA-Z]+$/),
    lastname: z.string().min(1).max(50).regex(/^[a-zA-Z]+$/).nullable(),
    surname: z.string().min(1).max(50).regex(/^[a-zA-Z]+$/).nullable(),
    username: z.string().min(1).max(50).regex(/^[A-Za-z][A-Za-z0-9_]*$/),

    email: z.string().min(5).max(50).regex(/^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/).nullable(),
    address: z.string().min(1).max(128).nullable(),
    phone: z.string().length(11).startsWith('+1').regex(/^[0-9]$/),
    password: z.string().min(8).max(128).regex(/^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*()\[\]{}<>?/|~_\-+]).+$/),

    level: z.int().nonnegative(),
    bonusBalance: z.int().nonnegative(),
    referralBalance: z.int().nonnegative(),

    preferences: z.record(z.any(), z.any()).nullable(),
    meta: z.record(z.any(), z.any()).nullable(),

    invitedBy: z.int().nullable(),
    invitedAt: z.date().nullable(),
    invitationLinks: z.array(z.string()),

    isBlocked: z.boolean(),
    blockedBy: z.int().nullable(),
    blockedAt: z.date().nullable(),
    blockingReason: z.string().max(128).nullable(),

    isDeleted: z.boolean(),
    deletedBy: z.int().nullable(),
    deletedAt: z.date().nullable(),
    deletionReason: z.string().max(128).nullable(),

    updatedBy: z.int().nullable(),
    updatedAt: z.date().nullable(),

    createdBy: z.int().nullable(),
    createdAt: z.date(),
}).strict();
```

steven-terrana committed:
add support mapping zod v4 string formats to fc supported generators

steven-terrana commented:
This schema _can_ be generated by the code as-is but it's incredible slow due to some fundamental issues with your UserSchema.

1. regex look aheads are not supported by fast check - see docs [1](https://fast-check.dev/blog/2023/05/30/whats-new-in-fast-check-3-10-0/#string-from-regex) [2](https://fast-check.dev/docs/core-blocks/arbitraries/combiners/string/#stringmatching) - solved this with a fc override that combines easier to generate substrings into a valid password. 
2. separating out length requirements from the regex results in a ton of strings being generated and then filtered until one matches - much faster to combine into the regex (also called out in the docs links above).
3. zod has an email built in in v4, so does fast check 
4. the phone number regex has a bug that should result it it never working because the regex expects a length of 1 but there's also a constraint of 11 characters. fixed the regex. 

<details>
<summary>expand here is a test file that suceeds for me updating your schema to something more compatible</summary>

``` typescript
import fc from 'fast-check'
import { z } from 'zod4'
import { Zod4FastCheck } from './zod-fast-check-module-proxy'

// Define schemas that need overrides as separate constants
const passwordSchema = z.string().min(8).max(128).regex(/^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*()\[\]{}<>?/|~_\-+]).+$/)

const userSchema = z.object({
  id: z.int(),

  roles: z.array(z.int()),

  // Use regex with length built-in for efficient generation
  firstname: z.string().regex(/^[a-zA-Z]{1,50}$/),
  lastname: z.string().regex(/^[a-zA-Z]{1,50}$/).nullable(),
  surname: z.string().regex(/^[a-zA-Z]{1,50}$/).nullable(),
  username: z.string().regex(/^[A-Za-z][A-Za-z0-9_]{0,49}$/),

  // Use Zod v4 built-in email format
  email: z.email().nullable(),
  address: z.string().min(1).max(128).nullable(),
  // Fixed phone regex: +1 followed by 9 digits = 11 chars total
  phone: z.string().regex(/^\+1[0-9]{9}$/),
  // Password with lookaheads - needs override
  password: passwordSchema,

  level: z.int().nonnegative(),
  bonusBalance: z.int().nonnegative(),
  referralBalance: z.int().nonnegative(),

  preferences: z.record(z.any(), z.any()).nullable(),
  meta: z.record(z.any(), z.any()).nullable(),

  invitedBy: z.int().nullable(),
  invitedAt: z.date().nullable(),
  invitationLinks: z.array(z.string()),

  isBlocked: z.boolean(),
  blockedBy: z.int().nullable(),
  blockedAt: z.date().nullable(),
  blockingReason: z.string().max(128).nullable(),

  isDeleted: z.boolean(),
  deletedBy: z.int().nullable(),
  deletedAt: z.date().nullable(),
  deletionReason: z.string().max(128).nullable(),

  updatedBy: z.int().nullable(),
  updatedAt: z.date().nullable(),

  createdBy: z.int().nullable(),
  createdAt: z.date(),
}).strict();

// Create arbitrary with override for password (lookaheads not supported by fc.stringMatching)
const userArbitrary = Zod4FastCheck()
  .override(passwordSchema,
    fc.tuple(
      fc.stringMatching(/^[a-z]{2,4}$/),   // lowercase
      fc.stringMatching(/^[A-Z]{2,4}$/),   // uppercase
      fc.stringMatching(/^[0-9]{2,4}$/),   // digits
      fc.constantFrom('!', '@', '#', '$', '%', '^', '&', '*')  // special
    ).map(([lower, upper, digits, special]) => `${lower}${upper}${digits}${special}`)
  )
  .inputOf(userSchema)

it('check user schema', () => {
  fc.assert(
    fc.property(userArbitrary, (user) => {
      const result = userSchema.safeParse(user)
      expect(result.success)
    }),
    { numRuns: 1000 }
  )
})
```

</details>

historydev commented:
The original password looked like this:

```javascript
password: z.string().min(8).max(128)
.regex(/[a-z]/)
.regex(/[A-Z]/)
.regex(/[0-9]/)
.regex(/[!@#$%^&*()\[\]{}<>?/|~_\-+]/),

```

I combined all the regular expressions into one because I thought the problem was in the splitting of the regular expression.

Yes, I read about the lookhead in your readme, thank you.

Regarding the length limit, yes, I encountered a frozen generation with the filter `length === X`. I solved the problem by adding the number of characters to the regular expression. I wouldn't want to use this in the zod itself; it's better to rewrite it for fast-check, because if there's a length-related error, I'll get invalid_format instead of too_small or too_big.

I'd like to control the email myself, because if something changes in the zod, it could cause a problem (although I don't plan to update without a compelling reason). I understand that there's a z.email, thanks.

historydev commented:
You are right, the generation time is reduced by ~3 times when use min, max in regex, thx u.

steven-terrana commented:
@DavidTimms - anything we can do to get this reviewed, merged, and published?

historydev commented:
maybe you publish own version on npm?

theredspoon commented:
[@traversable/zod-test](https://www.npmjs.com/package/@traversable/zod-test) may be a helpful replacement here. It supports Zod v4.
